File: python/tests/inference_test.py
"""Integration tests for run_inference."""

from dataclasses import dataclass

import pytest

from pedro_agentware.llm import Message, Role
from pedro_agentware.llm.request import ToolDefinition
from pedro_agentware.llm.response import Response, TokenUsage
from pedro_agentware.llm.response import ToolCall as LLMToolCall
from pedro_agentware.llmcontext.context_window import ContextWindowManager
from pedro_agentware.middleware.guardrails.error_tracker import ErrorTracker
from pedro_agentware.middleware.guardrails.response_validator import ResponseValidator
from pedro_agentware.middleware.guardrails.step_enforcer import StepEnforcer
from pedro_agentware.middleware.inference import (
    InferenceConfig,
    RetriesExhaustedError,
    run_inference,
)


@dataclass
class MockBackend:
    """Mock backend for testing."""

    responses: list[Response]
    call_count: int = 0

    def complete(self, messages: list[Message]) -> Response:
        """Return sequential mock responses."""
        if self.call_count >= len(self.responses):
            return self.responses[-1]
        resp = self.responses[self.call_count]
        self.call_count += 1
        return resp

    def supports_native_tool_calling(self) -> bool:
        return True

    def model_name(self) -> str:
        return "mock"

    def context_window_size(self) -> int:
        return 8192


class TestRunInference:
    """Tests for run_inference function."""

    @pytest.mark.asyncio
    async def test_successful_inference_with_tool_call(self):
        """Test successful inference returning tool calls."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        mock_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[mock_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)

        assert result is not None
        assert len(result.response.tool_calls) == 1
        assert result.response.tool_calls[0].name == "test_tool"
        assert result.attempts == 1

    @pytest.mark.asyncio
    async def test_successful_text_response_with_rescue(self):
        """Test text response successfully rescued to tool calls."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        mock_resp = Response(
            content='{"tool": "test_tool", "args": {"key": "value"}}',
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[mock_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=True)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)

        assert result is not None
        assert len(result.response.tool_calls) == 1

    @pytest.mark.asyncio
    async def test_retry_on_invalid_response(self):
        """Test retry logic on invalid response."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        invalid_resp = Response(
            content="This is just text, not a tool call",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        valid_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[invalid_resp, valid_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)

        assert result is not None
        assert result.attempts == 2

    @pytest.mark.asyncio
    async def test_retries_exhausted_error(self):
        """Test RetriesExhaustedError is raised when max attempts exceeded."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        invalid_resp = Response(
            content="This is just text, not a tool call",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[invalid_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=2,
        )

        with pytest.raises(RetriesExhaustedError):
            await run_inference(messages, cfg)

    @pytest.mark.asyncio
    async def test_context_compaction_triggered(self):
        """Test context compaction is triggered when needed."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        mock_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[mock_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        def custom_should_compact(messages: list[Message]) -> bool:
            return True

        ctx_manager = ContextWindowManager(context_window=1000)
        ctx_manager.should_compact = custom_should_compact
        ctx_manager.compact = lambda messages: messages[:1]

        cfg = InferenceConfig(
            client=backend,
            context_manager=ctx_manager,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)
        assert result is not None

    @pytest.mark.asyncio
    async def test_context_threshold_warning_injection(self):
        """Test context threshold warnings are injected."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        mock_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=8000),
        )

        backend = MockBackend(responses=[mock_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        ctx_manager = ContextWindowManager(context_window=10000)

        cfg = InferenceConfig(
            client=backend,
            context_manager=ctx_manager,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)
        assert result is not None

    @pytest.mark.asyncio
    async def test_error_tracker_records_failures(self):
        """Test error tracker is updated on validation failures."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        invalid_resp1 = Response(
            content="This is just text",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        invalid_resp2 = Response(
            content="Still not a tool call",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        valid_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[invalid_resp1, invalid_resp2, valid_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)
        error_tracker = ErrorTracker(max_errors_per_tool=5)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            error_tracker=error_tracker,
            max_attempts=5,
        )

        session_id = "test-session"

        result = await run_inference(messages, cfg, session_id=session_id)

        assert result is not None

    @pytest.mark.asyncio
    async def test_step_enforcer_blocks_premature_tools(self):
        """Test step enforcer blocks tools without prerequisites."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="final_tool",
            description="A final tool",
            input_schema={"type": "object", "properties": {}},
        )

        mock_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="final_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[mock_resp])
        validator = ResponseValidator(tool_names=["final_tool", "step1"], rescue_enabled=False)
        step_enforcer = StepEnforcer()
        step_enforcer.add_step("final_tool", prerequisites=["step1"])

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            step_enforcer=step_enforcer,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg, session_id="test-session")
        assert result is not None

    @pytest.mark.asyncio
    async def test_empty_response_triggers_retry(self):
        """Test empty response triggers retry."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        empty_resp = Response(
            content="",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        valid_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[empty_resp, valid_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)

        assert result is not None
        assert result.attempts == 2

    @pytest.mark.asyncio
    async def test_max_attempts_respects_config(self):
        """Test max attempts is respected from config."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        invalid_resp = Response(
            content="This is just text",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[invalid_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=5,
        )

        with pytest.raises(RetriesExhaustedError) as exc_info:
            await run_inference(messages, cfg)

        assert "5 attempts" in str(exc_info.value)

    @pytest.mark.asyncio
    async def test_default_max_attempts(self):
        """Test default max attempts is used when not specified."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        invalid_resp = Response(
            content="This is just text",
            tool_calls=[],
            finish_reason="stop",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[invalid_resp])
        validator = ResponseValidator(tool_names=["test_tool"], rescue_enabled=False)

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            validator=validator,
            max_attempts=0,
        )

        with pytest.raises(RetriesExhaustedError):
            await run_inference(messages, cfg)

    @pytest.mark.asyncio
    async def test_optional_guardrails_components(self):
        """Test inference works without optional guardrail components."""
        messages = [Message(role=Role.USER, content="test")]

        tool_def = ToolDefinition(
            name="test_tool",
            description="A test tool",
            input_schema={"type": "object", "properties": {}},
        )

        mock_resp = Response(
            content="",
            tool_calls=[LLMToolCall(id="1", name="test_tool", arguments={})],
            finish_reason="tool_calls",
            usage_tokens=TokenUsage(prompt_tokens=10, completion_tokens=5, total_tokens=15),
        )

        backend = MockBackend(responses=[mock_resp])

        cfg = InferenceConfig(
            client=backend,
            tool_specs=[tool_def],
            max_attempts=3,
        )

        result = await run_inference(messages, cfg)

        assert result is not None
        assert result.attempts == 1

File: typescript/tests/inference.test.ts
import type { Message, ToolDefinition } from "../src/llm/request.js";
import { Role } from "../src/llm/request.js";
import type { Response, ToolCall as LlmToolCall, TokenUsage } from "../src/llm/response.js";
import { ContextWindowManager } from "../src/llmcontext/context_window.js";
import { ResponseValidator } from "../src/middleware/guardrails/response_validator.js";
import { ErrorTracker } from "../src/middleware/guardrails/error_tracker.js";
import { StepEnforcer } from "../src/middleware/guardrails/step_enforcer.js";
import {
  InferenceConfig,
  RetriesExhaustedError,
  runInference,
} from "../src/middleware/inference.js";

interface MockBackend {
  complete(messages: Message[]): Response;
  supportsNativeToolCalling(): boolean;
  modelName(): string;
  contextWindowSize(): number;
}

function createMockResponse(
  content: string,
  toolCalls: LlmToolCall[] = [],
  usageTokens?: TokenUsage
): Response {
  return {
    content,
    tool_calls: toolCalls,
    finish_reason: toolCalls.length > 0 ? "tool_calls" : "stop",
    usage_tokens: usageTokens ?? {
      prompt_tokens: 10,
      completion_tokens: 5,
      total_tokens: 15,
    },
  };
}

function createToolDefinition(name: string): ToolDefinition {
  return {
    name,
    description: `A ${name} tool`,
    input_schema: { type: "object", properties: {} },
  };
}

describe("runInference", () => {
  let mockResponses: Response[];
  let callCount: number;

  beforeEach(() => {
    callCount = 0;
    mockResponses = [];
  });

  const createMockBackend = (): MockBackend => ({
    complete: (messages: Message[]): Response => {
      if (callCount >= mockResponses.length) {
        return mockResponses[mockResponses.length - 1];
      }
      const resp = mockResponses[callCount];
      callCount++;
      return resp;
    },
    supportsNativeToolCalling: () => true,
    modelName: () => "mock",
    contextWindowSize: () => 8192,
  });

  it("should return inference result on successful tool call", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("", [
        { id: "1", name: "test_tool", arguments: {} },
      ]),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);

    expect(result).not.toBeNull();
    expect(result!.response.tool_calls).toHaveLength(1);
    expect(result!.response.tool_calls[0].name).toBe("test_tool");
    expect(result!.attempts).toBe(1);
  });

  it("should rescue text response into tool calls", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse('{"tool": "test_tool", "args": {"key": "value"}}', []),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], true);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);

    expect(result).not.toBeNull();
    expect(result!.response.tool_calls).toHaveLength(1);
  });

  it("should retry on invalid response", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("This is just text, not a tool call", []),
      createMockResponse("", [{ id: "1", name: "test_tool", arguments: {} }]),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);

    expect(result).not.toBeNull();
    expect(result!.attempts).toBe(2);
  });

  it("should throw RetriesExhaustedError when max attempts exceeded", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("This is just text", []),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 2,
      stepIndex: 0,
    };

    await expect(runInference(messages, cfg)).rejects.toThrow(RetriesExhaustedError);
  });

  it("should trigger context compaction when needed", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("", [{ id: "1", name: "test_tool", arguments: {} }]),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);
    
    const mockCtxManager = {
      shouldCompact: () => true,
      compact: (msgs: Message[]) => [msgs[0]],
      addWarning: () => {},
      checkThresholds: () => undefined,
      updateTokenCount: () => {},
    } as unknown as ContextWindowManager;

    const cfg: InferenceConfig = {
      client: backend,
      contextManager: mockCtxManager,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);
    expect(result).not.toBeNull();
  });

  it("should inject context threshold warnings", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("", [{ id: "1", name: "test_tool", arguments: {} }], {
        prompt_tokens: 10,
        completion_tokens: 5,
        total_tokens: 8000,
      }),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);
    const ctxManager = new ContextWindowManager(10000);

    const cfg: InferenceConfig = {
      client: backend,
      contextManager: ctxManager,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);
    expect(result).not.toBeNull();
  });

  it("should use error tracker on validation failures", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("This is just text", []),
      createMockResponse("", [{ id: "1", name: "test_tool", arguments: {} }]),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);
    const errorTracker = new ErrorTracker(5);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      errorTracker,
      maxAttempts: 5,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg, "test-session");
    expect(result).not.toBeNull();
  });

  it("should check step enforcer for tool prerequisites", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("final_tool");

    mockResponses = [
      createMockResponse("", [{ id: "1", name: "final_tool", arguments: {} }]),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["final_tool", "step1"], false);
    const stepEnforcer = new StepEnforcer();
    stepEnforcer.addStep("final_tool", ["step1"]);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      stepEnforcer,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg, "test-session");
    expect(result).not.toBeNull();
  });

  it("should retry on empty response", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("", []),
      createMockResponse("", [{ id: "1", name: "test_tool", arguments: {} }]),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);

    expect(result).not.toBeNull();
    expect(result!.attempts).toBe(2);
  });

  it("should respect max attempts from config", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("This is just text", []),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 5,
      stepIndex: 0,
    };

    await expect(runInference(messages, cfg)).rejects.toThrow(
      "retries exhausted after 5 attempts"
    );
  });

  it("should use default max attempts when zero", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("This is just text", []),
    ];

    const backend = createMockBackend();
    const validator = new ResponseValidator(["test_tool"], false);

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      validator,
      maxAttempts: 0,
      stepIndex: 0,
    };

    await expect(runInference(messages, cfg)).rejects.toThrow(RetriesExhaustedError);
  });

  it("should work without optional guardrail components", async () => {
    const messages: Message[] = [{ role: Role.USER, content: "test" }];
    const toolDef = createToolDefinition("test_tool");

    mockResponses = [
      createMockResponse("", [{ id: "1", name: "test_tool", arguments: {} }]),
    ];

    const backend = createMockBackend();

    const cfg: InferenceConfig = {
      client: backend,
      toolSpecs: [toolDef],
      maxAttempts: 3,
      stepIndex: 0,
    };

    const result = await runInference(messages, cfg);

    expect(result).not.toBeNull();
    expect(result!.attempts).toBe(1);
  });
});